level-1/l1-lisp-threads.lisp

```diff
@@ -463,7 +463,35 @@
         (unless (zerop natural) natural)))))
 
 
-                         
+
+#+linux-target
+(defun set-os-thread-name (lisp-thread name)
+  "Set the OS-visible name of the thread with pthread_setname_np."
+  (let ((name
+          (if (<= (length name) 15)
+              name
+              (subseq name 0 15))))
+    (let ((natural (lisp-thread-os-thread lisp-thread)))
+      (when natural
+        (with-utf-8-cstr (name name)
```

Comment:
Note that (length name) is not, in general, guaranteed to be the same as (string-size-in-octets name :external-format :utf-8).

According to pthread_setname_np(3), it will return ERANGE if the string is too long. I suppose it's not the end of the world if the call just fails, but I'd like to see us ensure that the null-terminated encoded utf-8 data fits in 16 octets.

Reply:
Is string-size-in-octets usable at level 1? I tried introducing it but reloading the boot image fails with "Can't find symbol".

Reply:
You are right about string-size-in-octets.  However, you should be able to use ccl::utf-8-octets-in-string, which is loaded in level-0, and should be available.

If that's too much of a pain and you don't feel like dealing with the variable-length utf-8, you could use with-cstr and %get-cstring.  Those would use iso-8859-1.

I think this a nice little addition; I hope that you don't mind having to deal with my comments.

Reply:
Since I am interested in being able to spawn as many threads as possible, as quickly as possible, I would want to minimize the consing involved in thread creation. I.e. ideally for it to be constant time.

So I guess Linux guarantees you ASCII support, does it support a specific encoding like utf-8 or iso-8859-1 though? I can’t immediately tell from PTHREAD_SETNAME_NP(3). Maybe make a C string out of the first 15 ASCII characters, ideally without consing vita `subseq`?

Reply:
I've updated the pull request to use with-cstr and %get-cstring, so there should be no consing.